### PR TITLE
WIP: Experiments - FFI for Go

### PR DESCRIPTION
**Pull request by stefan-gorules**
Adds experimental FFI binding for GoLang.

🚧 Work in progress

### CODE REVIEW

**Comment by stefan-gorules** [comment]
TODO List before merge:
- Propagate errors to GoLang
- Write some tests for GoLang
- Set-up separate repository for keeping go files (as versioning requires tags and Go doesn't have package manager)
- Load testing to check for possible memory leaks

File: ffi-libs/go/bindings.h
Comment (ivanmiletic) [inline ffi-libs/go/bindings.h:1]: Move from ffi-libs/go to bindings/go

**Comment by stefan-gorules** [comment]
Status update:
- Bindings for Go have been completed on the high-level ✅ 
- Exposing errors to Go ✅ 
- First-draft of memory testing is done ✅ 

TODO:
- Figure out how we want to deploy bindings for Go (likely will be added as a separate repository)
- Polish out interfaces (prefixing C methods, allowing for evaluate_with_opts to pass struct instead of just trace: bool, etc.)
- Pipelines for Go (including testing)

**Comment by stefan-gorules** [comment]
For packer (releasing the bindings): https://github.com/egandro/zen-packer, will investigate to see how we want to use it to release C bindings (MVP will include go prefixed ones). The idea is to later include 1 more version that is more universal.

Perhaps we can use downcast_ref for the go-loader to avoid writing the code multiple times.

ty @egandro

**Comment by egandro** [comment]
Why not create a lib + h file?

It doesn't make sense to make this go exclusive, if you use Rust -> C bindings :)

Rename the bindings.h to zen_gorules.h (or something better) and also the library to libzengorules.a

(Check all packages in RPMs / DEBs for nameclashes - or you create a mess).

For go:

The issue with the -lpthread -lm needs to be adressed. I have to put it there in go with debian or I get linker errors.

**Comment by stefan-gorules** [comment]
@egandro

Yep, the plan is not to make it exclusive, but currently, there are some issues with binding cgo function handles with ZenEngine (needed for loaders to be definable).

I'm currently experimenting with the best way of handling multiple loaders through bindings, as we might need to define a specific loader interface for each language.

For example, for Go implementation, ZenEngine object needs to know handle_id (uint_ptr) in order to call a callback defined in Go. In C, we wouldn't need this and we would want a direct function callback instead. For other languages that we plan to support in future, there might be other needs.

This will likely be easiest to solve with Box<dyn Trait>, but I need to experiment a bit with it and see if it solves the issue.

**Comment by egandro** [comment]
> This will likely be easiest to solve with Box, but I need to experiment a bit with it and see if it solves the issue.

Ping me for eny assistance that you guys need.

**Comment by ivanmiletic** [comment]
We won't be directly supporting GO as there is an issue with uploading large files to GitHub that Go uses as package registry.

We will work next year on supporting GO via native OS bindings.

**Pull request closed without merging**